### ISSUE TITLE
Why is _iterator_deprecation_warning() impossible to suppress?

### ISSUE THREAD

**Issue by dreavjr** [NVIDIA/DALI#2140]
Does it mean that reader_name will very soon become mandatory? In that case, what is the design pattern to deal with ops.ExternalSource (especially in the case where num_outputs>1)?

If that is not the case (reader_name is not becoming mandatory, at least not very soon), I'd suggest the warning to have standard/expected Python behavior, making it possible to suppress.

**Comment by JanuszL** [comment]
Hi,
It is recommended to not set the Iterator size manually. If you do then you see the warning.
If you want to use ExternalSource don't set `size` of just use -1 `Providing -1 means that the iterator will work until StopIteration is raised from the inside of iter_setup()` and let the ExternalSource drive the framework iterator. In that case, providing the `size` doesn't make any sense as the user can use any custom logic determining epoch length.

**Comment by dreavjr** [comment]
Hi, thank you for the prompt answer!

I'm still a bit confused by what you mean. In the [template provided for using ExternalSource](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/frameworks/pytorch/pytorch-external_input.html) the logic is driven by using `StopIteration`, but `size`, `last_batch_padded`, and `fill_last_batch` are still set. Do you mean that in that example I could have omitted those parameters and the effect would've been the same?

**Comment by dreavjr** [comment]
I did some tests, I found out that the template provided would _almost_ work as intended (without need to set `size`) by changing the functions `__iter__` and `__next__` in `ExternalInputIterator` to:

```
def __iter__(self):
    self.i = 0
    self.count = 0
    shuffle(self.files)
    return self

def __next__(self):
    batch = []
    labels = []

    if self.count >= self.n:
        raise StopIteration

    for _ in range(self.batch_size):
        jpeg_filename, label = self.files[self.i].split(' ')
        f = open(self.images_dir + jpeg_filename, 'rb')
        batch.append(np.frombuffer(f.read(), dtype = np.uint8))
        labels.append(np.array([label], dtype = np.uint8))
        self.i = (self.i + 1) % self.n
        self.count += 1
    return (batch, labels)
```

With those changes the following line in the main routine...
```
pii = PyTorchIterator(pipe, size=eii.size, last_batch_padded=True, fill_last_batch=False)
```
...can become simply:
```
pii = PyTorchIterator(pipe)
```

But I say _almost_, because I could find no way to iterate over all items of the dataset exactly once per epoch without setting size=len(dataset), last_batch_padded=True, fill_last_batch=False. That is perhaps not critical for training, but it is for testing.

**Comment by JanuszL** [comment]
I'm sorry if you have been misled by this example. I just have updated it to be more precise https://github.com/NVIDIA/DALI/pull/2147. This is true that passing size, setting `last_batch_padded`, and `fill_last_batch` would yield the presented behavior.
My point is that the logic that the user may incorporate inside the `ExternalSource` may change epoch to epoch and providing the size once for all may not reflect the real data (two mentioned parameter relays on that).
The reason behind adding the `reader_name` parameter was to allow to drive the iterator directly from the reader - so when the data size is not equally divisible by the shard and batch size. shards differ from GPU to GPU, and are rotated epoch to epoch iterator can still correctly count the number of redundant in the last batch and treat them accordingly.
To sum up, you can still use the approach from the example but the warring is there to remind you that you need to know what you are doing. It is printed only during the iteration construction so it should not harm the speed to the code.

**Comment by dreavjr** [comment]
Thank you for the precision. I'm totally okay with the warning, I'm less okay with the fact it's enclosed inside a `with warnings.catch_warnings()` block that makes it uncatchable. IMHO, the call to warnings.warn() should be "naked", so the caller could decide how to treat it (e.g., to display it just once per run). As it is, the warning is mandatorily displayed every single time (it's an interface issue rather than a performance issue).

**Comment by JanuszL** [comment]
https://github.com/NVIDIA/DALI/pull/2154 should address your case.

**Comment by dreavjr** [comment]
Amazing, thank you! I think that closes this issue.
